iroh/examples/transfer.rs

```diff
@@ -0,0 +1,313 @@
+                    println!("node {node_id} disconnected with an error: {closed:#}");
+                }
+            })
+            .await;
+            if res.is_err() {
+                println!("node {node_id} did not disconnect within 3 seconds");
+            }
+            Ok::<_, anyhow::Error>(())
+        });
+    }
+
+    // stop with SIGINT (ctrl-c)
+    Ok(())
+}
+
+async fn fetch(ticket: &str, relay_url: Option<String>) -> anyhow::Result<()> {
+    let ticket: NodeTicket = ticket.parse()?;
+    let secret_key = SecretKey::generate();
+    let relay_mode = match relay_url {
+        Some(relay_url) => {
+            let relay_url = RelayUrl::from_str(&relay_url)?;
+            let relay_map = RelayMap::from_url(relay_url);
+            RelayMode::Custom(relay_map)
+        }
+        None => RelayMode::Default,
+    };
+    let endpoint = Endpoint::builder()
+        .secret_key(secret_key)
+        .alpns(vec![TRANSFER_ALPN.to_vec()])
+        .relay_mode(relay_mode)
+        .bind()
+        .await?;
+
+    let start = Instant::now();
+
+    let me = endpoint.node_id();
+    println!("node id: {me}");
+    println!("node listening addresses:");
+    for local_endpoint in endpoint
+        .direct_addresses()
+        .next()
+        .await
+        .context("no endpoints")?
+    {
+        println!("\t{}", local_endpoint.addr)
+    }
+
+    let relay_url = endpoint
+        .home_relay()
+        .expect("should be connected to a relay server, try calling `endpoint.local_endpoints()` or `endpoint.connect()` first, to ensure the endpoint has actually attempted a connection before checking for the connected relay server");
+    println!("node relay server url: {relay_url}\n");
+
+    // Attempt to connect, over the given ALPN.
+    // Returns a Quinn connection.
+    let conn = endpoint
+        .connect(ticket.node_addr().clone(), TRANSFER_ALPN)
+        .await?;
+    info!("connected");
+
+    // Use the Quinn API to send and recv content.
+    let (mut send, mut recv) = conn.open_bi().await?;
+
+    let message = format!("{me} is saying 'hello!'");
+    send.write_all(message.as_bytes()).await?;
+
+    // Call `finish` to close the send side of the connection gracefully.
+    send.finish()?;
+
+    let (len, ttfb, chnk) = drain_stream(&mut recv, false).await?;
+
+    // We received the last message: close all connections and allow for the close
+    // message to be sent.
+    endpoint.close(0u8.into(), b"bye").await?;
+
+    let duration = start.elapsed();
+    println!(
+        "Received {} B in {:.4}s with ttfb {}s in {} chunks",
+        HumanBytes(len as u64),
+        duration.as_secs_f64(),
+        ttfb.as_secs_f64(),
+        chnk
+    );
+    println!(
+        "Transferred {} in {:.4}, {}/s",
+        HumanBytes(len as u64),
+        duration.as_secs_f64(),
+        HumanBytes((len as f64 / duration.as_secs_f64()) as u64)
+    );
+
```

Comment:
You now need to explicitly call `endpoint.close().await` to make a best effort to let the other side know you closed the connection.  If you don't do this and the datagram carrying the close was lost the other side will trigger a timeout.  Awaiting the `endpoint.close` makes sure to wait until the datagram carrying the close is acknowledged (or times out on the acknowledgement).

Currently you probably also want to put this `endpoint.close().await` wrapped in a application-level timeout like you do on the other side because our API WRT timeouts is a bit crap right now...

Reply:
Something like now?

Reply:
Something like this here:

```suggestion
tokio::time::timeout(Duration::from_secs(3), async move {
    endpoint.close().await
})??;
```